### PR TITLE
Divide accounting metrics

### PR DESCRIPTION
**Pull request by vkatsuba**
Divide accounting metrics to `accounting`/`start`/`stop`/`update` types

### CODE REVIEW

File: src/eradius_server.erl
Comment (RoadRunnr) [inline src/eradius_server.erl:394]: this would that we potentially decode the request twice. That would seriously impact performance.

Please find a way to implement this without the decoding. Maybe pass the decoded request in this or do the counter for specific accounting request later.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: src/eradius_server.erl
Comment (vkatsuba) [reply src/eradius_server.erl]: Done

File: src/eradius_server.erl
Comment (RoadRunnr) [inline src/eradius_server.erl:397]: I don't see why this has to use a list (but I might be missing something here).

If you leave the inc_{reply,request}_counter for accreq/resp in place, you would only have to deal with the Status-Type here and you could use a plain match on the Status-Type.
Also, calling inc_request_counter_accounting/inc_response_counter_accounting from inc_request_counter(accreq, _)/inc_reply_counter(accresp, _) might further simplify the logic.

File: src/eradius_server.erl
Comment (RoadRunnr) [inline src/eradius_server.erl:418]: `{_, ?RStatus_Type, _, _, _}` is a #attribute record. So matching with `#attribute{id = ...}` should work.

Also this whole list logic could be rewritten with a ets:match_spec_run/2 on the list of attributes.

Additional bonus points will be given for caching the compile match_spec in persistent_term ;-)

File: test/eradius_metrics_SUITE.erl
Comment (RoadRunnr) [inline test/eradius_metrics_SUITE.erl:135]: this is suspicious, why would the accounting request accounted twice?

Comment (vkatsuba) [reply test/eradius_metrics_SUITE.erl:135]: Looks like because we have `accountRequests` and `accountRequestsStart` metrics and as total of those metrics is `2`. Eg: we have `accountRequestsStart: 1` and `accountRequestsStop: 1` and also we keep all `accountRequest` and `accountRequest` will be `2` and total accounting metrics of those points will be `4`.

File: src/eradius_server.erl
Comment (vkatsuba) [reply src/eradius_server.erl]: Done.

File: src/eradius_server.erl
Comment (vkatsuba) [reply src/eradius_server.erl]: If I correct got your point - Done.

File: src/eradius_server.erl
Comment (vkatsuba) [inline src/eradius_server.erl:426]: Looks like, this code can be refactored and move to macros:
```erlang
-define(REQ_ACC_MATCH_SPEC_COMPILE, ets:match_spec_compile(ets:fun2ms(fun
    ({#attribute{id = ?RStatus_Type}, ?RStatus_Type_Start})  -> {accounting_start, accountRequestsStart};
    ({#attribute{id = ?RStatus_Type}, ?RStatus_Type_Stop})   -> {accounting_stop, accountRequestsStop};
    ({#attribute{id = ?RStatus_Type}, ?RStatus_Type_Update}) -> {accounting_update, accountRequestsUpdate} end))).

-define(RESP_ACC_MATCH_SPEC_COMPILE, ets:match_spec_compile(ets:fun2ms(fun
    ({#attribute{id = ?RStatus_Type}, ?RStatus_Type_Start})  -> {accounting_start, accountResponsesStart};
    ({#attribute{id = ?RStatus_Type}, ?RStatus_Type_Stop})   -> {accounting_stop, accountResponsesStop};
    ({#attribute{id = ?RStatus_Type}, ?RStatus_Type_Update}) -> {accounting_update, accountResponsesUpdate} end))).
```

Comment (RoadRunnr) [reply src/eradius_server.erl:426]: that would rerun ets:match_spec_compile every time the macro is used, i.e. every time inc_response_counter_accounting is executed. ms_transform.hrl will only optimize fun2ms, the match_spec_compile. By putting the result of match_spec_compile into persistent term, the compile is executed only once for the first invocation.

Comment (vkatsuba) [reply src/eradius_server.erl:426]: Thanks.

File: test/eradius_metrics_SUITE.erl
Comment (vkatsuba) [reply test/eradius_metrics_SUITE.erl]: Fixed

File: METRICS.md
Comment (ebengt) [inline METRICS.md:24]: Prometheus likes counters to have suffix _total, like replies_total. Perhaps accounting_requests_start_total is a good name? Similar for the other new counters.

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (ebengt) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:23]: I think you can remove ''of'' from the description. Same for the other new counters.

File: include/eradius_lib.hrl
Comment (ebengt) [inline include/eradius_lib.hrl:101]: Nothing wrong, just wondering if there is a way to define the same record members once, instead of multiple times.

File: src/eradius_client.erl
Comment (ebengt) [inline src/eradius_client.erl:37]: Perhaps it is possible to do a separate commit with white space changes?

File: src/eradius_counter.erl
Comment (ebengt) [inline src/eradius_counter.erl:220]: Only white space changes?

File: src/eradius_client.erl
Comment (ebengt) [inline src/eradius_client.erl:445]: What happens if HAVE_PERSISTENT_TERM is not defined?
No client_request_counter_account_match_spec_compile/0 at all?

**Review by ebengt — COMMENTED** [review]
Looks good, apart from the -ifdef that might leave the code without a function => fail to compile.
But I could have misunderstood something.

File: METRICS.md
Comment (vkatsuba) [reply METRICS.md:24]: I was take a look to the `client_accounting_responses` counter and then made a decision create without `_total` prefix, I not sure that without suffix `_total`  the Prometheus will work worse. But if `counter` without suffix `_tota` looks like confusing I will change it.

File: src/eradius_client.erl
Comment (vkatsuba) [reply src/eradius_client.erl:37]: This was done by `atom` editor

File: src/eradius_client.erl
Comment (vkatsuba) [reply src/eradius_client.erl:445]: Will be used `client_request_counter_account_match_spec_compile/0` after `else` block without using `persistent_term`. Only for Erlang 21.

File: src/eradius_counter.erl
Comment (vkatsuba) [reply src/eradius_counter.erl:220]: Not only white space changes, was added also :
```
accountRequestsStart, accountRequestsStop, accountResponsesUpdate, accountResponsesStart, accountResponsesStop, accountRequestsUpdate %% For #nas_counter{}
% AND
accountRequestsStart, accountRequestsStop, accountResponsesUpdate, accountResponsesStart, accountResponsesStop, accountRequestsUpdate %% For #client_counter{}

```

File: include/eradius_lib.hrl
Comment (0xAX) [reply include/eradius_lib.hrl:101]: @ebengt yes, probably it would be good idea to merge records into one, but in separate PR as this one will bring more unrelated changes

File: src/eradius_counter.erl
Comment (0xAX) [reply src/eradius_counter.erl:220]: Could you please just add new clauses to the end, in other way diff touches all the lines and it is hard to review what really was changed

File: METRICS.md
Comment (0xAX) [reply METRICS.md:24]: @ebengt yes, you're right here, but let's make this in separate PR as well, because as @vkatsuba mentioned we also should rename a couple of metrics that are not related to this PR

Comment (ebengt) [reply METRICS.md:24]: So first introduce new counter client_accounting_responses here. Then make a new pull request to change the name to client_accounting_responses_total?

File: src/eradius_client.erl
Comment (ebengt) [reply src/eradius_client.erl:37]: It is okay (for me) to let atom editor move white space around. But please, in a separate pull request.

File: METRICS.md
Comment (RoadRunnr) [reply METRICS.md:24]: the real question is why is that a separate metric to begin with. Should the accounting type be a tag on accounting_requests_total?

Comment (vkatsuba) [reply METRICS.md]: Add labels into `accounting_requests_total`. @ebengt, @RoadRunnr, @0xAX, please review.

File: METRICS.md
Comment (0xAX) [inline METRICS.md:64]: as you've added total to the `client_accounting_request`, add `total` here too

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (0xAX) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:100]: do we really need lists:flatten here?

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (0xAX) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:105]: in what case it could return this?

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (0xAX) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:259]: add spaces here, so nas_id will be right after under server_name

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (0xAX) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:255]: What if we will call `fetch_metric` without certain accounting_type to get all accounting_requests_total or accounting_responses_total for all accounting types?

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (0xAX) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:297]: There is a warning that there is no new line in the end of file, it is not really critical but other files have it

**Comment by 0xAX** [comment]
Besides small comments and the fact that we should be able to fetch all 3 types of accounting_requests_total and other accounting metrics without labels - LGTM

File: include/eradius_lib.hrl
Comment (vkatsuba) [inline include/eradius_lib.hrl:71]: Remove it.

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (RoadRunnr) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:268]: checking type against ACCT_TYPES is a bit excessive. `acct_type` is either set and valid, or not set. Checking the return for `proplists:get_value` for undefined should be sufficient.

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (RoadRunnr) [inline applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:290]: checking type against ACCT_TYPES is a bit excessive. acct_type is either set and valid, or not set at all. Checking the return for proplists:get_value for undefined should be sufficient.

**Review by RoadRunnr — COMMENTED** [review]
two smallish comments, but looks good overall.

File: applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl
Comment (vkatsuba) [reply applications/eradius_prometheus_collector/src/eradius_prometheus_collector.erl:268]: I add this checking by reason that the type can be put incorrectly(please, see test point `ok = check_metric_multi({bad_type, EradiusRequestType}, client_accounting_requests_total, [{server_name, good}, {acct_type, bad_type}], 0)`).

**Review by 0xAX — APPROVED** [review]
LGTM

**Review by RoadRunnr — APPROVED** [review]

**Comment by 0xAX** [comment]
@RoadRunnr I'm going to merge this PR if you're ok with everything

**Pull request merged**
